### PR TITLE
Calculate and assign projection for SemiReorderRule

### PR DESCRIPTION
**Pull request by Youngwb**
#2796 
1. Calculate and assign projection for SemiReorderRule
2. Fix PhysicalHashAggregateOperator equals error

### CODE REVIEW

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentTest.java [hunk 1]
```diff
@@ -5337,7 +5337,54 @@ public void testEqDoubleCast() throws Exception {
     public void testNotEqStringCast() throws Exception {
         String sql = "select 'a' != v1 from t0";
         String plan = getFragmentPlan(sql);
-        System.out.println(plan);
         Assert.assertTrue(plan.contains("CAST(1: v1 AS VARCHAR(1048576)) != 'a'\n"));
     }
+
+    @Test
+    public void testSemiJoinReorder() throws Exception {
```
[the remark this answers is not in the document]
Comment (Youngwb) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentTest.java:5344]: Yes

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java [hunk 2]
```diff
@@ -82,6 +87,26 @@ public boolean check(final OptExpression input, OptimizerContext context) {
...
+        OptExpression leftChildJoinRightChild = input.inputAt(0).inputAt(1);
+        ColumnRefSet leftChildJoinRightChildOutputColumns = leftChildJoinRightChild.getOutputColumns();
+
+        Projection leftChildJoinProjection = leftChildJoin.getProjection();
+        HashMap<ColumnRefOperator, ScalarOperator> rightExpression = new HashMap<>();
+        HashMap<ColumnRefOperator, ScalarOperator> semiExpression = new HashMap<>();
+        if (leftChildJoinProjection != null) {
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : leftChildJoinProjection.getColumnRefMap()
+                    .entrySet()) {
+                if (!entry.getValue().isColumnRef() &&
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java:100]: `!entry.getValue().isColumnRef()`  is common

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java [hunk 3]
```diff
@@ -94,12 +119,47 @@ public boolean check(final OptExpression input, OptimizerContext context) {
...
+                        .collect(Collectors.toMap(Function.identity(), Function.identity()));
+            } else {
+                expressionProject = Maps.newHashMap(leftChildJoinRightChild.getOp().getProjection().getColumnRefMap());
+            }
+
+            // Use leftChildJoinRightChild projection to rewrite the rightExpression, it's like two project node merge.
+            ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(expressionProject);
+            Map<ColumnRefOperator, ScalarOperator> rewriteMap = Maps.newHashMap(expressionProject);
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : rightExpression.entrySet()) {
+                rewriteMap.put(entry.getKey(), entry.getValue().accept(rewriter, null));
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java:149]: need clone here?

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java:149]: No need, rightExpression will not used

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java [hunk 4]
```diff
@@ -104,23 +104,23 @@ public boolean isUseStreamingPreAgg() {
...
     }
 
     @Override
-    public boolean equals(Object obj) {
-        if (!(obj instanceof PhysicalHashAggregateOperator)) {
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
```
[the remark this answers is not in the document]
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java:115]: It just follow other operator equals, I think we should use one style

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java (hunk 2 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java [hunk 5]
```diff
@@ -94,12 +119,47 @@ public boolean check(final OptExpression input, OptimizerContext context) {
                     .collect(Collectors.toMap(Function.identity(), Function.identity()));
         }
 
-        LogicalJoinOperator newSemiJoin = new LogicalJoinOperator.Builder().withOperator(topJoin)
-                .setProjection(new Projection(projectMap)).build();
+        LogicalJoinOperator newSemiJoin;
+        // build new semi join projection
+        if (semiExpression.isEmpty()) {
+            newSemiJoin = new LogicalJoinOperator.Builder().withOperator(topJoin)
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java:126]: maybe new expression can't be assign in newSemiJoin

**Review by HangyuanLiu — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**